packages/firestore/test/integration/api/snasphot_listener_source.test.ts

```diff
@@ -0,0 +1,778 @@
+            { k: 'b', sort: 1 },
+            { k: 'a', sort: 0 }
+          ]);
+
+          // Un-listen then re-listen to the limit query.
+          limitUnlisten();
+          limitUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'asc'), limit(2)),
+            { source: ListenSource.Cache },
+            storeLimitEvent.storeEvent
+          );
+          snapshot = await storeLimitEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'a', sort: 0 },
+            { k: 'b', sort: 1 }
+          ]);
+          expect(snapshot.metadata.fromCache).to.equal(true);
+
+          // Add a document that would change the result set.
+          await addDoc(coll, { k: 'd', sort: -1 });
+
+          // Verify both queries get expected results.
+          snapshot = await storeLimitEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'd', sort: -1 },
+            { k: 'a', sort: 0 }
+          ]);
+          expect(snapshot.metadata.hasPendingWrites).to.equal(true);
+
+          snapshot = await storeLimitToLastEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'a', sort: 0 },
+            { k: 'd', sort: -1 }
+          ]);
+          expect(snapshot.metadata.hasPendingWrites).to.equal(true);
+
+          // Un-listen to limitToLast, update a doc, then re-listen limitToLast.
+          limitToLastUnlisten();
+
+          await updateDoc(doc(coll, 'a'), { k: 'a', sort: -2 });
+          limitToLastUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'desc'), limitToLast(2)),
+            { source: ListenSource.Cache },
+            storeLimitToLastEvent.storeEvent
+          );
+
+          // Verify both queries get expected results.
+          snapshot = await storeLimitEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'a', sort: -2 },
+            { k: 'd', sort: -1 }
+          ]);
+          expect(snapshot.metadata.hasPendingWrites).to.equal(true);
+
+          snapshot = await storeLimitToLastEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'd', sort: -1 },
+            { k: 'a', sort: -2 }
+          ]);
+          // We listened to LimitToLast query after the doc update.
+          expect(snapshot.metadata.hasPendingWrites).to.equal(false);
+
+          limitUnlisten();
+          limitToLastUnlisten();
+        });
+      });
+
+      it('can listen to default source first and then cache', () => {
+        const testDocs = {
+          a: { k: 'a', sort: 0 }
+        };
+        return withTestCollection(persistence, testDocs, async coll => {
+          // Listen to the query with default options, which will also populates the cache
+          const storeDefaultEvent = new EventsAccumulator<QuerySnapshot>();
+          const defaultUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'asc')),
```

Comment:
Can we add a where clause here, and do multiple documents. OrderBy with only one document don't really show the difference.